src/Icicle/Source/Checker/Function.hs

```diff
@@ -59,10 +60,15 @@ checkF' fun env cons
  = do -- Give each argument a fresh type variable
       env' <- foldM bindArg env $ arguments fun
       -- Get the annotated body
-      (q',_,cons')  <- generateQ (body fun) env' cons
+      (q', subs, cons')  <- generateQ (body fun) env' cons
+
+      -- Look up the argument types after solving all constraints.
+      -- Because they started as fresh unification variables,
+      -- they will end up being unified to the actual types.
+      (args, cs) <- List.unzip <$> mapM (lookupArg subs env' cons') (arguments fun)
 
       -- Find all leftover constraints and nub them
-      let constrs = ordNub $ fmap snd cons'
+      let constrs = ordNub $ fmap snd (concat cs)
```

Comment:
ah got it. I don't think there will be any constraints attached to the arguments, because they're bound as non-functions. this is ok though.